### ISSUE TITLE
Non-deterministically receiving too many answers from rules with value predicates

### ISSUE THREAD

**Issue by matterpale** [vaticle/typedb#6349]
## Description

When given a schema with rules about two related attributes and some test data, only Workbase infers the attributes correctly. Console / client-java read transaction returns unexpected values (and once it does, Workbase shows them, too).

## Environment

1. OS: Docker on Win10
2. TypeDB version: TypeDB 2.1.1
3. TypeDB client: console (or client-java)
4. Workbase version: 2.1.0

## Reproducible Steps

Steps to create the smallest reproducible scenario:
1. Create schema and data gql files:
s.gql -
```
define

name sub attribute, value string;

volume sub attribute, value long;

timestamp sub attribute, value datetime;

available sub attribute, value boolean;

deliverable sub attribute, value boolean;

Customer sub entity,
    owns name @key,
    plays Order:customer;

Resource sub entity,
    owns name,
    owns volume;

Ingredient sub entity,
    owns name,
    owns volume,
    owns available,
    plays Product:ingredient;

Order sub relation,
    owns name @key,
    owns timestamp,
    owns deliverable,
    relates customer,
    relates product;

Product sub relation,
    owns name @key,
    plays Order:product,
    relates ingredient;

rule ingredient-available-rule:
when {
    $i isa Ingredient, has name $n, has volume $iv;
    $r isa Resource, has name $n, has volume $rv;
    $rv > $iv;
} then {
    $i has available true;
};

rule ingredient-unavailable-rule:
when {
    $i isa Ingredient, has name $n, has volume $iv;
    not {
        $r isa Resource, has name $n, has volume $rv;
        $rv > $iv;
    };
} then {
    $i has available false;
};

rule undeliverable-rule:
when {
    $o (product: $p) isa Order;
    $p (ingredient: $i) isa Product;
    $i isa Ingredient, has available false;
} then {
    $o has deliverable false;
};

rule deliverable-rule:
when {
    $o (product: $p) isa Order;
    $p isa Product;
    not {
        $i isa Ingredient;
        $p (ingredient: $i);
        $i has available false;
    };
} then {
    $o has deliverable true;
};
```

d.gql -
```
insert

$i1 isa Ingredient,
    has name "Liquid 1",
    has volume 120;
    
$i2 isa Ingredient,
    has name "Liquid 2",
    has volume 30;
    
$i3 isa Ingredient,
    has name "Liquid 3",
    has volume 250;

$r11 isa Resource,
    has name "Liquid 1",
    has volume 100;

$r12 isa Resource,
    has name "Liquid 1",
    has volume 500;
    
$r21 isa Resource,
    has name "Liquid 2",
    has volume 200;
    
$r22 isa Resource,
    has name "Liquid 2",
    has volume 250;

$r31 isa Resource,
    has name "Liquid 3",
    has volume 10;

$r32 isa Resource,
    has name "Liquid 3",
    has volume 220;

$p1 (ingredient: $i1, ingredient: $i2) isa Product,
    has name "Product 1";

$p2 (ingredient: $i1, ingredient: $i3) isa Product,
    has name "Product 2";

$john isa Customer,
    has name "John Doe";

$o1 (product: $p1, customer: $john) isa Order,
    has name "Order 1";

$o2 (product: $p2, customer: $john) isa Order,
    has name "Order 2";

$o3 (product: $p1, customer: $john) isa Order,
    has name "Order 3";
```

2. Move these files to your Docker container:opt/typedb-all-linux/

3. Run these commands in console (either as a script or one by one):
```
database create test
transaction test schema write
    source s.gql
    commit
transaction test data write
    source d.gql
    commit
```

4. Open TypeDB Workbase (v2.1.0) connect to your server, select database "test" and run:
`match $x isa Order, has deliverable $d;`
(notice correct inference)

5. Go back to console, open a read transaction with inference on and check the same thing (optionally add the name attribute for readability):
```
transaction test data read --infer true
test::data::read> match $x isa Order, has deliverable $d, has name $n;
```
(notice incorrect inference, Order 2 is both deliverable and undeliverable)

6. Go back to TypeDB Workbase, run the same query as in Step 4 and notice the same incorrect inference as in Step 5.

## Expected Output

The output given by Workbase in Step 4 matches expectations

## Actual Output

Console and client-java infer that Order 2 is both deliverable and undeliverable.

## Additional information

![Screenshot1](https://user-images.githubusercontent.com/56684558/120484165-17c0b680-c3b3-11eb-9837-ad9746465b89.png)
![Screenshot2](https://user-images.githubusercontent.com/56684558/120484172-18f1e380-c3b3-11eb-94a6-de8fe94113be.png)
![Screenshot3](https://user-images.githubusercontent.com/56684558/120484179-1b543d80-c3b3-11eb-8db3-03e7322a7c19.png)

**Comment by matter-pale** [comment]
I should note that when trying to come up with the smallest reproducible scenario, I first omitted the `ingredient-available-rule` and `ingredient-unavailable-rule` and I set the ingredients `available` values directly. However, that did not result in the described misbehavior. Thus, the problem is likely related to the fact that `deliverable` is determined by inference and depends on `available` that **is also determined by inference**.

Why Workbase is the only way to get correct answers, I have no clue. However, both console and especially client-java consistently give the described incorrect answers. That is unless there is a flaw in the underlying schema (which again would not explain the difference between reasoning results).

**Comment by jmsfltchr** [comment]
Thanks for writing such a beautiful issue @matter-pale! Workbase executes additional queries/ ConceptAPI calls in the background so they may be the cause of different behaviour there (I can't say for sure). I have been able to reproduce the same results as you have reported via the console.

From what I can see, `Order 2` should be undeliverable, but it is both deliverable and undeliverable. This combined with your comment above suggests that the rule `rule ingredient-available-rule` has set ingredient 3 to available when it shouldn't have.

**Comment by matter-pale** [comment]
Thanks for your time @jmsfltchr !
Glad you could reproduce the issue. It does seem like Ingredient 3 has been mistakenly set to available, however, no query returns that statement to me.

**Comment by matter-pale** [comment]
With my experience with fiddling with the bug and my very limited knowledge of what is happening under the hood, I would suspect one of two things:
Either that when evaluating `deliverable-rule` and coming across the line with `available` attribute, the reasoner "notices" it can be set by another rule and at that point under some circumstances jumps to the wrong conclusion.
Or there might be an issue with evaluating the negation block (or perhaps with the way it is written). Maybe I'm just shooting in the dark, though.
Anyway, let me know if I can be of further assistance (I could share my pojos to test it with the client-java or something..)

**Comment by jmsfltchr** [comment]
@matter-pale yes it's something along those lines! Actually if you'd like to dig in then you could use the visual debugging. To do so you can run the query having opened the transaction with `--infer true --parallel false --trace-inference true`. First check that the bug is still present (sometimes they disappear when running in non-parallel), then you should find a set of `.dot` files in your server logs directory. There is one `.dot` file per answer, and they are numbered in the same order that you see the answers. You can convert them to sag using `graphviz` and inspect the internal message-passing in the reasoner! So then you could check the one for the incorrect answer and verify that the Resolvers that handle the condition and conclusion for the `ingredient-available-rule` rule gives a green arrow (indicating an answer) when it shouldn't. :)

**Comment by matter-pale** [comment]
It did actually disappear when running in non-parallel...

**Comment by dmitrii-ubskii** [comment]
Appears to have been fixed as of TypeDB 2.13:
```
tmp::data::read> match $x isa Order, has deliverable $d, has name $name;
{
    $d true isa deliverable;
    $name "Order 1" isa name;
    $x iid 0x847080028000000000000001 (product: iid 0x847080018000000000000000, customer: iid 0x826e80018000000000000000) isa Order;
}
{
    $d true isa deliverable;
    $name "Order 3" isa name;
    $x iid 0x847080028000000000000002 (product: iid 0x847080018000000000000000, customer: iid 0x826e80018000000000000000) isa Order;
}
{
    $d false isa deliverable;
    $name "Order 2" isa name;
    $x iid 0x847080028000000000000000 (product: iid 0x847080018000000000000001, customer: iid 0x826e80018000000000000000) isa Order;
}
```
